onnx/defs/parser.h

```diff
@@ -0,0 +1,382 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Experimental language syntax and parser for ONNX. Please note that the syntax as formalized
+// by this parser is preliminary and may change.
+
+#pragma once
+
+#include <ctype.h>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+
+#include "onnx/onnx_pb.h"
+
+#include "onnx/common/status.h"
+#include "onnx/string_utils.h"
+
+namespace ONNX_NAMESPACE {
+
+using namespace ONNX_NAMESPACE::Common;
+
+using IdList = google::protobuf::RepeatedPtrField<std::string>;
+
+using NodeList = google::protobuf::RepeatedPtrField<NodeProto>;
+
+using AttrList = google::protobuf::RepeatedPtrField<AttributeProto>;
+
+using ValueInfoList = google::protobuf::RepeatedPtrField<ValueInfoProto>;
+
+#define PARSE_ERROR(...)                                                                                        \
+  do {                                                                                                          \
+    return Status(                                                                                              \
+        NONE, FAIL, ONNX_NAMESPACE::MakeString("[ParseError at position ", GetCurrentPos(), "]", __VA_ARGS__)); \
+  } while (0)
+
+#define PARSER_CHECK(cond, msg) \
+  if (!(cond))                  \
+  PARSE_ERROR(msg)
+
+#define CHECK_PARSER_STATUS(status) \
```

Comment:
I understand that usage of this macro lets you avoid the if-statements at the call site but this is not a very modern-cpp approach. You could even switch to exceptions instead.

Reply:
Yes, but ONNX also supports a build-option to build without exceptions. All things considered, it seems better to avoid exceptions for parser-errors.